### 01. Importing Libraries

In [1]:
# import libraries
import pandas as pd
import numpy as np
import os

### 02. Importing dataset

In [2]:
#create a path
path = r'C:\Users\Lumex\Documents\Achievement_4_Project\InstaCart Basket Analysis'
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge.pkl'))

### 03. Creating a subset for the dataframe 

In [3]:
# creating a subset of the dataframe because of the large dataset.
df = ords_prods_merge[:1000000]

In [4]:
# checking the subset
df.shape

(1000000, 16)

### 04. User-defined function

In [5]:
# defining a function with a if statement
def price_label(row):

  if row['prices'] <= 5:
    return 'Low-range product'
  elif (row['prices'] > 5) and (row['prices'] <= 15):
    return 'Mid-range product'
  elif row['prices'] > 15:
    return 'High range'
  else: return 'Not enough data'

In [6]:
#Apply the function
df['price_range'] = df.apply(price_label, axis=1)

C:\Users\Lumex\AppData\Local\Temp\ipykernel_6440\3959542531.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis=1)


In [7]:
df['price_range'].value_counts(dropna = False)

price_range
Mid-range product    673340
Low-range product    314392
High range            12268
Name: count, dtype: int64

In [8]:
df['prices'].max()

25.0

### 05. if statement with loc() function

In [9]:
#if statement with loc
df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'

C:\Users\Lumex\AppData\Local\Temp\ipykernel_6440\888299827.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'


In [10]:
df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product' 

In [11]:
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [12]:
df['price_range'].value_counts(dropna = False)

price_range
Mid-range product    673340
Low-range product    314392
High range            12268
Name: count, dtype: int64

### 06. If statement with for loop

In [13]:
# what is the slowest and busiest day
result = []

for value in ords_prods_merge["orders_day_of_week"]:
  if value == 0:
    result.append("Busiest day")
  elif value == 4:
    result.append("Least busy")
  else:
    result.append("Regularly busy")

In [14]:
result

['Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy

In [15]:
# create a new column within your ords_prods_merge dataframe
ords_prods_merge['busiest_day'] = result

In [16]:
ords_prods_merge['busiest_day'].value_counts(dropna = False)

busiest_day
Regularly busy    22416875
Busiest day        6204182
Least busy         3783802
Name: count, dtype: int64

### 07. Suppose your clients have changed their minds about the labels you created in your “busiest_day” column

In [17]:
# check the days that are busiest and slowest.
ords_prods_merge['orders_day_of_week'].value_counts(dropna = False)

orders_day_of_week
0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: count, dtype: int64

In [18]:
# create label for the busiest_days and slowest_days?
result = []
for value in ords_prods_merge["orders_day_of_week"]:
  if value == 0 or value == 1:
    result.append("Busiest days")
  elif value == 3 or value == 4:
    result.append("Slowest days")
  else:
    result.append("Regularly busy")

In [19]:
result

['Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest day

In [20]:
# create a new column within your ords_prods_merge dataframe
ords_prods_merge['busiest_days'] = result

In [21]:
# check the value count for the busiest days
ords_prods_merge['busiest_days'].value_counts(dropna = False)

busiest_days
Regularly busy    12916111
Busiest days      11864412
Slowest days       7624336
Name: count, dtype: int64

In [22]:
# confirm that both value count in new columns are equal
#busiest_day column
ords_prods_merge['busiest_day'].shape

(32404859,)

In [23]:
#busiest_days column
ords_prods_merge['busiest_days'].shape

(32404859,)

#### Observation about the new column: The new column created contain same number of 32,404,859 rows which means all the rows were allocated. The differnces is just in the number of allocation.

### 08. When too many users make Instacart orders at the same time, the app freezes. The senior technical officer at Instacart wants you to identify the busiest hours of the day.

In [24]:
# viewing the column in the dataframe
ords_prods_merge.head()

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0           0   2539329        1             1                   2   
1           0   2539329        1             1                   2   
2           0   2539329        1             1                   2   
3           0   2539329        1             1                   2   
4           0   2539329        1             1                   2   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     NaN         196                  1   
1                  8                     NaN       14084                  2   
2                  8                     NaN       12427                  3   
3                  8                     NaN       26088                  4   
4                  8                     NaN       26405                  5   

   reordered _merge                             product_name  aisle_id  \
0          0   both                                     Soda        77   
1          0   both  Organic Unsweetened Vanilla Almond Milk        91   
2          0   both                      Original Beef Jerky        23   
3          0   both               Aged White Cheddar Popcorn        23   
4          0   both         XL Pick-A-Size Paper Towel Rolls        54   

   department_id  prices match     busiest_day    busiest_days  
0              7     9.0  both  Regularly busy  Regularly busy  
1             16    12.5  both  Regularly busy  Regularly busy  
2             19     4.4  both  Regularly busy  Regularly busy  
3             19     4.7  both  Regularly busy  Regularly busy  
4             17     1.0  both  Regularly busy  Regularly busy

In [25]:
ords_prods_merge.tail()

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
32404854     3421081   2977660   206209            13                   1   
32404855     3421081   2977660   206209            13                   1   
32404856     3421081   2977660   206209            13                   1   
32404857     3421081   2977660   206209            13                   1   
32404858     3421081   2977660   206209            13                   1   

          order_hour_of_day  days_since_prior_order  product_id  \
32404854                 12                     7.0       14197   
32404855                 12                     7.0       38730   
32404856                 12                     7.0       31477   
32404857                 12                     7.0        6567   
32404858                 12                     7.0       22920   

          add_to_cart_order  reordered _merge  \
32404854                  5          1   both   
32404855                  6          0   both   
32404856                  7          0   both   
32404857                  8          0   both   
32404858                  9          0   both   

                                   product_name  aisle_id  department_id  \
32404854                           Tomato Paste         9              9   
32404855        Brownie Crunch High Protein Bar         3             19   
32404856  High Protein Bar Chunky Peanut Butter         3             19   
32404857    Chocolate Peanut Butter Protein Bar         3             19   
32404858    Roasted & Salted Shelled Pistachios       117             19   

          prices match     busiest_day  busiest_days  
32404854     5.6  both  Regularly busy  Busiest days  
32404855     5.9  both  Regularly busy  Busiest days  
32404856     4.2  both  Regularly busy  Busiest days  
32404857     4.9  both  Regularly busy  Busiest days  
32404858     1.7  both  Regularly busy  Busiest days

In [26]:
# Check value counts in "order_hour_of_day" column.
ords_prods_merge['order_hour_of_day'].value_counts()

order_hour_of_day
10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: count, dtype: int64

In [27]:
# Create if statement with for loops labeling the hour of the day
#the hour of the day will be subdivided into three which will be most order, fewest orders, average.
busiest_hour = []
for value in ords_prods_merge["order_hour_of_day"]:
  if value in [10, 11, 14, 15, 13, 12, 16, 9]: busiest_hour.append("Most orders")
  elif value in [23, 6, 0, 1, 5, 2, 4, 3]: busiest_hour.append("Fewest orders")
  else: busiest_hour.append("Average orders")

In [28]:
busiest_hour

['Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Avera

In [29]:
# Create new column "busiest_period_of_day" in ords_prods_merge.
ords_prods_merge['busiest_period_of_day'] = busiest_hour

In [30]:
#view the new column in the ord_prods_merge dataframe
ords_prods_merge.head()

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0           0   2539329        1             1                   2   
1           0   2539329        1             1                   2   
2           0   2539329        1             1                   2   
3           0   2539329        1             1                   2   
4           0   2539329        1             1                   2   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     NaN         196                  1   
1                  8                     NaN       14084                  2   
2                  8                     NaN       12427                  3   
3                  8                     NaN       26088                  4   
4                  8                     NaN       26405                  5   

   reordered _merge                             product_name  aisle_id  \
0          0   both                                     Soda        77   
1          0   both  Organic Unsweetened Vanilla Almond Milk        91   
2          0   both                      Original Beef Jerky        23   
3          0   both               Aged White Cheddar Popcorn        23   
4          0   both         XL Pick-A-Size Paper Towel Rolls        54   

   department_id  prices match     busiest_day    busiest_days  \
0              7     9.0  both  Regularly busy  Regularly busy   
1             16    12.5  both  Regularly busy  Regularly busy   
2             19     4.4  both  Regularly busy  Regularly busy   
3             19     4.7  both  Regularly busy  Regularly busy   
4             17     1.0  both  Regularly busy  Regularly busy   

  busiest_period_of_day  
0        Average orders  
1        Average orders  
2        Average orders  
3        Average orders  
4        Average orders

#### Print the frequency of the new column

In [31]:
# Print value counts in "busiest_period_of_day" column.
ords_prods_merge['busiest_period_of_day'].value_counts(dropna = False)

busiest_period_of_day
Most orders       21118071
Average orders     9997651
Fewest orders      1289137
Name: count, dtype: int64

#### Export the dataframe as pickle

In [32]:
ords_prods_merge.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merged_derived.pkl'))